### PR TITLE
[CDAP-3718] Add metadata cleaner in the DeletedProgramHandlerStage.

### PR DESCRIPTION
**Pull request by henry-cask**
As part of Application update, need to delete all metadata related to no longer exist program in an application.

This is done in ```DeletedProgramHandlerStage.process```

### CODE REVIEW

**Comment by henry-cask** [comment]
Build:
http://builds.cask.co/browse/CDAP-RBT468-1

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java [hunk 1]
```diff
@@ -377,6 +378,35 @@ public void testInvalidTags() throws IOException {
...
+
+    // Get properties from deleted (flow) program - should return 404
+    getPropertiesFromInvalidEntity(program);
+
+    //Delete the App after stopping the flow
+    org.apache.http.HttpResponse response =
+      doDelete(getVersionedAPIPath("apps/WordCountApp/", Constants.Gateway.API_VERSION_3_TOKEN,
+                                   TEST_NAMESPACE1));
+    Assert.assertEquals(200, response.getStatusLine().getStatusCode());
+    response = doDelete(getVersionedAPIPath("apps/WordCountApp/", Constants.Gateway.API_VERSION_3_TOKEN,
```
Comment (bdmogal) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: there should be a deleteApp method?

File: cdap-examples/Purchase/src/main/java/co/cask/cdap/examples/purchase/PurchaseApp.java [hunk 2]
```diff
@@ -47,7 +47,7 @@ public void configure() {
     createDataset("userProfiles", KeyValueTable.class);
 
     // Process events in realtime using a Flow
-    addFlow(new PurchaseFlow());
+    //addFlow(new PurchaseFlow());
```
Comment (bdmogal) [inline cdap-examples/Purchase/src/main/java/co/cask/cdap/examples/purchase/PurchaseApp.java]: unintended change?

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java (hunk 1 above)
Comment (henry-cask) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: Yeah, I was copy-paste from other test class. Will use deleteApp instead.

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java [hunk 3]
```diff
@@ -377,6 +378,30 @@ public void testInvalidTags() throws IOException {
...
+    Map<String, String> properties = getProperties(program);
+    Assert.assertEquals(2, properties.size());
+
+    // Deploy WordCount App without Flow program.
+    deploy(WordCountMinusFlowApp.class, Constants.Gateway.API_VERSION_3_TOKEN, TEST_NAMESPACE1);
+
+    // Get properties from deleted (flow) program - should return 404
+    getPropertiesFromInvalidEntity(program);
+
+    //Delete the App after stopping the flow
```
Comment (gokulavasan) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: Where are we starting/stopping the flow?

**Comment by gokulavasan** [comment]
One minor nit. Rest LGTM :+1:

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java (hunk 3 above)
Comment (henry-cask) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: deploying an app does not mean start the flow, does it?

Comment (bdmogal) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: Nope, you have to start it explicitly. Also, if you try deleting app while the flow is started, you should get a 403/409 (not sure)

Comment (henry-cask) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: Yeah, I do not need to start/stop the flow. Just need the updated app to contain no Flow program.

Comment (bdmogal) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: yup, just update the comment.

**Comment by bdmogal** [comment]
@henry-cask has this been tested on a cluster? If so, build is green, please merge as soon as you get a chance.

**Comment by henry-cask** [comment]
@bdmogal yeah it has been tested in cluster

**Comment by henry-cask** [comment]
merging ...

**Comment by henry-cask** [comment]
Pass build: http://builds.cask.co/browse/CDAP-RBT468-3

**Pull request merged**